api/org/commcare/api/persistence/SqlIndexedStorageUtility.java

```diff
@@ -0,0 +1,418 @@
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+
+        }
+
+        return null;
+    }
+
+    @Override
+    public T getRecordForValue(String fieldName, Object value)
+            throws NoSuchElementException, InvalidIndexException {
+        Connection c = null;
+        PreparedStatement preparedStatement = null;
+        try {
+            c = this.getConnection();
+            preparedStatement =
+                    SqlHelper.prepareTableSelectStatement(c, this.tableName,
+                            new String[]{fieldName}, new String[]{(String)value},
+                            prototype.newInstance());
+            if (preparedStatement == null) {
+                throw new NoSuchElementException();
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+            if (!rs.next()) {
+                throw new NoSuchElementException();
+            }
+            byte[] mBytes = rs.getBytes(org.commcare.modern.database.DatabaseHelper.DATA_COL);
+            return readFromBytes(mBytes);
+        } catch (SQLException | InstantiationException |
+                IllegalAccessException | NullPointerException e) {
+            e.printStackTrace();
+        } finally {
+            try {
+                if (c != null) {
+                    c.close();
+                }
+                if (preparedStatement != null) {
+                    preparedStatement.close();
+                }
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+        }
+        return null;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#add(org.javarosa.core.util.externalizable.Externalizable)
+     */
+    public int add(T e) throws StorageFullException {
+        this.write(e);
+        return 1;
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#close()
+     */
+    public void close() {
+        // TODO Auto-generated method stub
+
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#destroy()
+     */
+    public void destroy() {
+        // TODO Auto-generated method stub
+
+    }
+
+    /* (non-Javadoc)
+     * @see org.javarosa.core.services.storage.IStorageUtility#exists(int)
+     */
+    public boolean exists(int id) {
+        PreparedStatement preparedStatement = null;
+        Connection c = null;
+        try {
+            c = getConnection();
+            preparedStatement = SqlHelper.prepareIdSelectStatement(c, this.tableName, id);
+            if (preparedStatement == null) {
+                return false;
+            }
+            ResultSet rs = preparedStatement.executeQuery();
+            if (rs.next()) {
+                return true;
+            }
+        } catch (Exception e) {
+            System.out.println("SqlIndexedStorageUtility readBytes exception: " + e);
```

Comment:
Log instead of sysout? Errr, or am I mistaken: do we have non-persistent logging at the commcare repo level?